### ISSUE TITLE
Letter case for npm packages

### ISSUE THREAD

**Issue by scanossmining** [package-url/purl-spec#136]
Hello,

According to the specifications for the npm purl type, the package name inside the purl should be lowercased, however this would trigger some purl duplication. npmjs.com hosts case sensitive packages, some examples being:

- base64 and Base64
- dva, DVA, DvA, dVa

**Comment by stevespringett** [comment]
By convention, npm is lowercase. The spec states the name SHOULD be lowercased, but it doesn't state that it MUST be lowercased. I do think we need clarity here though.

This issue was recently discovered in the official Package URL Java implementation and has subsequently been corrected. https://github.com/package-url/packageurl-java/pull/38 There may be other implementations also affected.

**Comment by scanossmining** [comment]
Hi @stevespringett , thanks for answering this, it's clearer now. 
Before creating the issue, I've read the npm type specs here https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#npm ,where it says 

> Per the package.json spec, new package "must not have uppercase letters in the name", therefore the must be lowercased.

I guess it's a typo there and the ending was supposed to say "therefore the **package** must be lowercased" ?

**Comment by pombredanne** [comment]
@scanossmining re:
> I guess it's a typo there and the ending was supposed to say "therefore the package must be lowercased" ?

Yes, good catch! PR welcomed!

**Comment by iamwillbar** [comment]
From the [npm documentation](https://docs.npmjs.com/cli/v7/configuring-npm/package-json):

```
The name must be less than or equal to 214 characters. This includes the scope for scoped packages.
The names of scoped packages can begin with a dot or an underscore. This is not permitted without a scope.
New packages must not have uppercase letters in the name.
The name ends up being part of a URL, an argument on the command line, and a folder name. Therefore, the name can't contain any non-URL-safe characters.
```

The use of the word "New" on the third rule implies that this doesn't apply to existing packages. Since we want purls to be able to reference "new" and "old" packages that means we'd need to support mixed-case. The case should match the case used in the package's `package.json`.

**Comment by wesleytodd** [comment]
> The use of the word "New" on the third rule implies that this doesn't apply to existing packages.

Sorry for reviving an old thread, but you are correct that there are older packages with names not following current rules.

I will expand on this to say that while I agree with some of the apparent goals of this project, you are going to have a really tough time trying to get adoption in the JS ecosystem because you chose to break from the existing well understood and widely deployed formats. Honestly looking at the work in this repo, it looks like y'all skipped the step of asking folks who work on the JS package managers what they think.

Not sure if y'all are too far along here to make a change in direction, but I would guess based on this issue that if you want a standard which is used by the npm registry that you will really need to re-assess the current proposal.

**Comment by pombredanne** [comment]
This came up again. @wesleytodd if you can reel in npm maintainers that would be awesome.

Here is the status so far:

People are technically correct that npm used to support mixed case names in the past. But this is has long been not the case.
The npm package.json spec is clear on this. Please see: https://docs.npmjs.com/cli/v11/configuring-npm/package-json

> New packages must not have uppercase letters in the name. 

New was 10+ years ago:
https://web.archive.org/web/20150516132705/https://docs.npmjs.com/files/package.json

There are possibly a few legacy old warts alright, and @zeke did nicely track them all at https://github.com/nice-registry/mixed-case-package-names and 719 legacy npm names have **overlapping names**, for which both a mixed-case and lowercase package exist.


Questions:
1. Is it allowed to publish an npm package to the public registry with mixed case names today?
2. Are there any active, non-deprecated packages with mixed case on the public registry? 


@matt-phylum ping because of your comment in:
- https://github.com/package-url/purl-spec/pull/614#discussion_r2309985147



Overall, I am fine going either way, but I would not want to get weird validation just as a exception for a few old, legacy packages from 10+ years ago that are likely unused. So it would be great to get evidence.

**Comment by pombredanne** [comment]
@isaacs @wraithgar your take you be much welcomed on the topic!

**Comment by matt-phylum** [comment]
I don't think it matters how long it has been since npmjs.com allowed new packages with mixed case.
- Packages with mixed case names still exist and still can be downloaded from npmjs.com, but cannot be named by the current PURL spec.
- NPM is clearly case sensitive. You cannot install a lowercase package by providing an uppercase name. The registry API is case sensitive and the package manager does not search for packages with alternative capitalization.
- npmjs.com is not the only NPM package registry. Somebody can set up their own registry where all packages are uploaded with mixed case names and none of those packages could be named by PURL because the normalization rules throw away data that is required by the NPM registry in order to resolve the correct package.

This can make PURL unsuitable for SBOMs. For example, if somebody controls MEAN (an application erroneously published as a library?) they can publish a new, undesirable version `pkg:npm/MEAN@1.0.1`. If PURL is used for naming the packages in an SBOM, the dependency would be replaced with `pkg:npm/mean@1.0.1` (a pointless but harmless library). If you search SBOMs for `pkg:npm/MEAN@1.0.1`, you won't find it because the normalization will have made it look like `pkg:npm/mean@1.0.1`.

**Comment by isaacs** [comment]
This is correct, npm is technically not case-insensitive, and if the PURL spec is forcing package names to be lowercase, then that is going to lose information that is almost certainly going to be relevant.

Initially, npm did not prevent the publishing of packages with uppercase letters in the name. This oversight was rectified relatively quickly (within 1-2 years of npm's existence, if I remember correctly), but not before several uppercase names were published. That of course causes problems on case-insensitive systems (typically macOS and Windows), but at the time, removing these packages from the registry would have also caused problems, hence the compromise of not allowing uppercase letters in new packages published to the registry.

Ie, `pkg:npm/MEAN@1.0.1` and `pkg:npm/mean@1.0.1` are fundamentally different things, even though they cannot be installed side by side on many systems due to case-insensitive file systems.

**Comment by pombredanne** [comment]
@isaacs Thanks, I think this nails it. Also, I got report that private registries allow mixed case packages. Let's fix these all in the spec and in tools and libraries :)

**Comment by conorfitch** [comment]
Any update on changing npm's `case_sensitive` to True @pombredanne?
I was also wondering if the `case_sensitive` of Go and Ruby should be changed to True, because the case is preserved on their registries.

**Comment by matt-phylum** [comment]
NPM and Go* and Ruby are case sensitive, not just case preserving. If you attempt to install a package with the wrong capitalization, you will not get the expected package.

\* Go is always case sensitive when using Go modules, which is the normal case today. AFAIK Go was also case sensitive before modules, but you could often get away with using the wrong case because of the way packages were resolved.
